pkg/proc/native/proc_freebsd.go

```diff
@@ -0,0 +1,368 @@
+		}
+
+		if status.Exited() {
+			dbp.postExit()
+			return nil, proc.ErrProcessExited{Pid: wpid, Status: status.ExitStatus()}
+		}
+
+		tid, pl_flags, _, err := ptraceGetLwpInfo(wpid)
+		if err != nil {
+			return nil, fmt.Errorf("ptraceGetLwpInfo err %s %d", err, pid)
+		}
+		th, ok := dbp.threads[tid]
+		if ok {
+			th.Status = (*WaitStatus)(status)
+		}
+
+		if status.StopSignal() == sys.SIGTRAP {
+			if pl_flags&sys.PL_FLAG_EXITED != 0 {
+				delete(dbp.threads, tid)
+				PtraceCont(tid, 0)
+				continue
+			} else if pl_flags&sys.PL_FLAG_BORN != 0 {
+				th, err = dbp.addThread(int(tid), false)
+				if err != nil {
+					if err == sys.ESRCH {
+						// process died while we were adding it
+						continue
+					}
+					return nil, err
+				}
+				if halt {
+					return nil, nil
+				}
+				if err = th.Continue(); err != nil {
+					if err == sys.ESRCH {
+						// thread died while we were adding it
+						delete(dbp.threads, int(tid))
+						continue
+					}
+					return nil, fmt.Errorf("could not continue new thread %d %s", tid, err)
+				}
+				continue
+			}
+		}
+
+		if th == nil {
+			continue
+		}
+
+		if (halt && status.StopSignal() == sys.SIGSTOP) || (status.StopSignal() == sys.SIGTRAP) {
+			return th, nil
+		}
+
+		// TODO(dp) alert user about unexpected signals here.
+		if err := th.resumeWithSig(int(status.StopSignal())); err != nil {
+			if err == sys.ESRCH {
+				return nil, proc.ErrProcessExited{Pid: dbp.pid}
+			}
+			return nil, err
+		}
+	}
+}
+
+func (dbp *Process) loadProcessInformation() {
+}
+
+// Helper function used here and in threads_freebsd.go
+// Return the status code
+func status(pid int) rune {
+	status := rune(C.find_status(C.int(pid)))
+	return status
+}
+
+// Used by stop and singleStep
+// waitFast is like wait but does not handle process-exit correctly
+func (dbp *Process) waitFast(pid int) (int, *sys.WaitStatus, error) {
+	var s sys.WaitStatus
+	wpid, err := sys.Wait4(pid, &s, 0, nil)
+	return wpid, &s, err
+}
+
+// Only used in this file
+func (dbp *Process) wait(pid, options int) (int, *sys.WaitStatus, error) {
+	var s sys.WaitStatus
+	wpid, err := sys.Wait4(pid, &s, options, nil)
+	return wpid, &s, err
+}
+
+// Only used in this file
+func (dbp *Process) exitGuard(err error) error {
+	if err != sys.ESRCH {
+		return err
+	}
+	if status(dbp.pid) == StatusZombie {
+		_, err := dbp.trapWaitInternal(-1, false)
+		return err
+	}
+
+	return err
+}
+
+// Used by ContinueOnce
+func (dbp *Process) resume() error {
+	// all threads stopped over a breakpoint are made to step over it
+	for _, thread := range dbp.threads {
+		if thread.CurrentBreakpoint.Breakpoint != nil {
+			if err := thread.StepInstruction(); err != nil {
+				return err
+			}
+			thread.CurrentBreakpoint.Clear()
+		}
+	}
+	// all threads are resumed
+	PtraceCont(dbp.pid, 0)
+	return nil
+}
+
+// Used by ContinueOnce
+// stop stops all running threads and sets breakpoints
+func (dbp *Process) stop(trapthread *Thread) (err error) {
```

Comment:
How do threads other than the trapthread get stopped? Since you aren't callling Thread.stop here.

Reply:
In FreeBSD, whenever one thread is trapped, the entire process (all threads) are stopped. Since we call "stop" only after "trapWait" (in ContinueOnce), all threads are always stopped, so no need to explicitly check for that.